src/index.ts

```diff
@@ -99,6 +158,81 @@ export interface RegisterOptions extends CreateOptions {
   preferTsExts?: boolean | null
 }
 
+/*
+ * This interface exists solely for generating a JSON schema for tsconfig.json.
+ * We do *not* extend the compiler's tsconfig interface.  Instead we handle that
+ * on a schema level, via "allOf", so we pull in the same schema that VSCode
+ * already uses.
+ */
+/**
+ * tsconfig schema which includes "ts-node" options.
+ * @allOf [{"$ref": "https://schemastore.azurewebsites.net/schemas/json/tsconfig.json"}]
+ */
+export interface TsConfigSchema {
+  /**
+   * ts-node options.  See also: https://github.com/TypeStrong/ts-node#configuration-options
+   *
+   * ts-node offers TypeScript execution and REPL for node.js, with source map support.
+   */
+  'ts-node': TsConfigOptions
+}
+
+export interface TsConfigOptions
+  extends Omit<RegisterOptions,
+    | 'transformers'
+    | 'readFile'
+    | 'fileExists'
+    | 'skipProject'
+    | 'project'
+    | 'dir'
+  > {
+  requires?: Array<string>
+}
+
+/**
+ * Helper for obtaining ts-node options from multiple sources and merging them.
+ * @internal
+ */
+export class OptionsHelper {
+  constructor (fields: Partial<OptionsHelper>) {
+    Object.assign(this, fields)
+  }
+
+  /**
+   * Options specified explicitly via --flags or options object passed to ts-node's public API.
+   * They should override everything else.
+   */
+  public explicitOptions: RegisterOptions = {}
+
+  /**
+   * Default options that are overridden by all other sources. Env vars fall into this category.
+   */
+  public defaultOptions: RegisterOptions = DEFAULTS
+
+  /** Options pulled from tsconfig.json */
+  public tsconfigOptions: RegisterOptions = {}
+
+  merge (): RegisterOptions {
+    return defaults(this.defaultOptions, this.tsconfigOptions, this.explicitOptions)
+  }
+}
+
+/**
+ * Like Object.assign or splatting, but never overwrites with `undefined`.
+ * This matches the behavior for argument and destructuring defaults.
+ * @internal
+ */
+export function defaults<T> (...sources: Array<T>): T {
+  const merged: any = {}
+  for (const source of sources) {
+    for (const key of Object.keys(source)) {
+      const value = (source as any)[key]
+      if (value !== undefined) merged[key] = value
```

Comment:
Probably want to do `!= null` since some options can be `null`.

Reply:
Do we want a `null` to be able to override a non-null value specified elsewhere?  The idea was `undefined` means the option is unspecified, whereas `null` means we explicitly want to override the default with `null`.  I was mimicking the JS language behavior of defaults in arguments and destructuring.  Will that cause problems?

Reply:
> Will that cause problems?

No, I think your reasoning sounds solid. I hadn't considered someone would use `null`, it's historically just an input because of using the `yn` library which returns `null` when a value is neither true or false.